Prevent unit tests from modifying local Badger settings

ghostwords opened the pull request:
Following up on #1391.

cowlicks commented on src/js/background.js:
```diff
@@ -391,13 +391,13 @@ Badger.prototype = {
+    migrationLevel: 0,
     seenComic: false,
-    migrationLevel: 0
+    showCounter: true,
```
was this just alphabetizing stuff?

ghostwords replied:
Yep.

cowlicks commented on src/tests/lib/qunit_config.js:
```diff
@@ -1 +1,41 @@
-QUnit.config.testTimeout = 6400;
+/* globals badger:false */
```
Could this be moved to `src/tests/qunit_config.js` or `src/tests/tests/qunit_config.js`? When I see a "`lib`" I unusually think of it as only containing code from a third party.

ghostwords replied:
I'm (slowly) moving the third-party stuff into `lib/vendor`. I was thinking `lib` (vs. `js`) could be our own library/non-main-entry-point type files.

Ghoughpteighbteau commented on src/tests/lib/qunit_config.js:
```diff
@@ -1 +1,41 @@
+    // set defaults
+    var settings = badger.storage.getBadgerStorageObject("settings_map");
+    for (let key in badger.defaultSettings) {
+      settings.setItem(key, badger.defaultSettings[key]);
```
ah nice.

If privacy badger forms an opinion about example.com, could that mess up the unit tests? Could there be entries in the users action_map that effect the unit tests? Seems like something to trip on as well.

cowlicks replied:
That is a good point.

Maybe we could also override `chrome.storage.local.get` so when we initialize PB in the test page, it will start with default data. What do you think @ghostwords ?

cowlicks commented on src/tests/lib/qunit_config.js:
```diff
@@ -1 +1,41 @@
+  // runs before any tests begin
+  QUnit.begin(() => {
+    // disable storage persistence
+    chrome.storage.local.set = () => {};
```
Originally I was concerned that this instance of `chrome.storage.local.set` was shared by the tests, and the real privacy badger, so overriding this would effect the real PB. But this isn't the case (!!).

ghostwords replied:
I think when the unit tests page is open, there are two Badgers each with its own memory/execution context, but storage is shared.

cowlicks commented on src/tests/lib/qunit_config.js:
```diff
@@ -1 +1,41 @@
+  QUnit.done(() => {
+  });
+
+  QUnit.testStart(() => {
```
Does this happen for each test? Could we just do this once in the beginning?

ghostwords replied:
No, we need to reset state between tests. See how several tests fail if you move this to `QUnit.begin()`.

ghostwords replied:
I should clarify the comments in these two methods I think.

cowlicks commented on src/tests/lib/qunit_config.js:
```diff
@@ -1 +1,41 @@
+    // restore original settings and heuristic learning
+    ['action_map', 'settings_map', 'snitch_map'].forEach(item => {
+      let obj = badger.storage.getBadgerStorageObject(item);
+      obj.updateObject(BACKUP[item]);
```
Since `chrome.storage.local.set` is overridden, this will not actually go to disk. Is the purpose of this to restore the in-memory testing PB to its original state?

Should we also do this with other stateful in memory objects like the `badger.tabData`^

ghostwords replied:
Maybe, I'll leave a note for later.

ghostwords commented:
@cowlicks @Ghoughpteighbteau OK, I changed how storage resetting works (all storage gets reset, not just `settings_map`), worked around a newly surfaced race condition (`dnt_hashes` get set via an XHR to `eff.org` and so might not get populated in time for tests that use them), and added notes for later. Let me know if you see anything we should fix now.

Ghoughpteighbteau commented:
nice. 

I think ideally, tests should be stating their assumptions about global settings, but this is practical, and doesn't preclude that. Unit tests are much safer now. 

lgtm

cowlicks commented on src/tests/lib/qunit_config.js:
```diff
@@ -6,21 +6,22 @@
+      cb({
+        // don't open the firstrun page though
+        settings_map: {
+          isFirstRun: false,
```
👍

cowlicks commented:
Nice work! lgtm

ghostwords commented:
Cool, thanks all for reviewing!
